browser/brave_rewards/checkout_dialog.cc

```diff
@@ -0,0 +1,343 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_rewards/checkout_dialog.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/values.h"
+#include "brave/browser/brave_rewards/rewards_service_factory.h"
+#include "brave/common/webui_url_constants.h"
+#include "brave/components/brave_rewards/browser/rewards_service.h"
+#include "brave/components/brave_rewards/browser/rewards_service_impl.h"
+#include "brave/components/brave_rewards/common/constants.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/webui/constrained_web_dialog_ui.h"
+#include "components/payments/content/payment_request.h"
+#include "components/payments/core/payer_data.h"
+#include "components/web_modal/web_contents_modal_dialog_host.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/web_contents.h"
+#include "third_party/blink/public/mojom/payments/payment_request.mojom.h"
+#include "ui/web_dialogs/web_dialog_delegate.h"
+
+using content::WebContents;
+using content::WebUIMessageHandler;
+using payments::PaymentRequest;
+using payments::mojom::PaymentErrorReason;
+
+namespace brave_rewards {
+
+namespace {
+
+constexpr int kDialogWidth = 548;
+constexpr int kDialogMinHeight = 200;
+constexpr int kDialogMaxHeight = 800;
+
+void OnSKUProcessed(base::WeakPtr<payments::PaymentRequest> request,
+                    const ledger::type::Result result,
+                    const std::string& value) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  if (!request) {
+    return;
+  }
+
+  if (result == ledger::type::Result::LEDGER_OK) {
+    payments::mojom::PaymentDetailsPtr details =
+        payments::mojom::PaymentDetails::New();
+    details->id = value;
+    request->spec()->UpdateWith(std::move(details));
+    request->Pay();
+    return;
+  }
+  request->OnError(PaymentErrorReason::UNKNOWN, errors::kBatTransactionFailed);
+}
+
+void OnGetPublisherDetailsCallback(base::WeakPtr<PaymentRequest> request,
+                                   WebContents* contents,
```

Comment:
we'd better extract `contents` from the request

Reply:
resolved 👍